subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathCompareStrategy.java

```diff
@@ -41,56 +43,129 @@ private ClasspathCompareStrategy() {
-        Iterator<Map.Entry<String, FileSystemLocationFingerprint>> currentEntries = currentSnapshots.entrySet().iterator();
-        Iterator<Map.Entry<String, FileSystemLocationFingerprint>> previousEntries = previousSnapshots.entrySet().iterator();
-        while (true) {
-            if (currentEntries.hasNext()) {
-                Map.Entry<String, FileSystemLocationFingerprint> current = currentEntries.next();
+        TrackingVisitor trackingVisitor = new TrackingVisitor(visitor);
+        ChangeState changeState = new ChangeState(propertyTitle, includeAdded, trackingVisitor, currentSnapshots, previousSnapshots);
+
+        while (trackingVisitor.isConsumeMore() && changeState.hasMoreToProcess()) {
+            changeState.processChange();
+        }
+        return trackingVisitor.isConsumeMore();
+    }
+
+    private static class TrackingVisitor implements ChangeVisitor {
+        private final ChangeVisitor visitor;
+        private boolean consumeMore = true;
+
+        private TrackingVisitor(ChangeVisitor visitor) {
+            this.visitor = visitor;
+        }
+
+        @Override
+        public boolean visitChange(Change change) {
+            if (consumeMore) {
+                consumeMore = visitor.visitChange(change);
+            }
+            return consumeMore;
+        }
+
+        public boolean isConsumeMore() {
+            return consumeMore;
+        }
+    }
+
+    private static class ChangeState {
+        private Map.Entry<String, FileSystemLocationFingerprint> current;
+        private Map.Entry<String, FileSystemLocationFingerprint> previous;
+        private final boolean includeAdded;
+        private final ChangeVisitor changeConsumer;
+        private final Iterator<Map.Entry<String, FileSystemLocationFingerprint>> currentEntries;
+        private final Map<String, FileSystemLocationFingerprint> currentSnapshots;
+        private final Iterator<Map.Entry<String, FileSystemLocationFingerprint>> previousEntries;
+        private final Map<String, FileSystemLocationFingerprint> previousSnapshots;
+        private final String propertyTitle;
+
+        private ChangeState(String propertyTitle, boolean includeAdded, ChangeVisitor changeConsumer, Map<String, FileSystemLocationFingerprint> currentSnapshots, Map<String, FileSystemLocationFingerprint> previousSnapshots) {
+            this.propertyTitle = propertyTitle;
+            this.includeAdded = includeAdded;
+            this.changeConsumer = changeConsumer;
+            this.currentEntries = currentSnapshots.entrySet().iterator();
+            this.currentSnapshots = currentSnapshots;
+            this.previousEntries = previousSnapshots.entrySet().iterator();
+            this.previousSnapshots = previousSnapshots;
+            this.current = nextEntry(currentEntries);
+            this.previous = nextEntry(previousEntries);
+        }
+
+        void processChange() {
```

Comment:
Since this is one of those cases where we need to make sure everything is handled, I would argue that we should consider the no-early-`return` variant:

```suggestion
        void processChange() {
            if (current == null) {
                if (previous != null) {
                    removed();
                }
            } else if (previous == null) {
                added();
            } else {
                FileSystemLocationFingerprint currentFingerprint = current.getValue();
                FileSystemLocationFingerprint previousFingerprint = previous.getValue();
                String currentNormalizedPath = currentFingerprint.getNormalizedPath();
                String previousNormalizedPath = previousFingerprint.getNormalizedPath();
                if (!currentNormalizedPath.equals(previousNormalizedPath)) {
                    removed();
                    added();
                } else if (currentFingerprint.getNormalizedContentHash().equals(previousFingerprint.getNormalizedContentHash())) {
                    current = nextEntry(currentEntries);
                    previous = nextEntry(previousEntries);
                } else if (currentNormalizedPath.isEmpty()) {
                    String currentAbsolutePath = current.getKey();
                    String previousAbsolutePath = previous.getKey();
                    if (!currentAbsolutePath.equals(previousAbsolutePath)) {
                        if (!currentSnapshots.containsKey(previousAbsolutePath)) {
                            removed();
                        } else if (!previousSnapshots.containsKey(currentAbsolutePath)) {
                            added();
                        } else {
                            removed();
                            added();
                        }
                    } else {
                        modified();
                    }
                } else {
                    modified();
                }
            }
        }
```

WDYT?